### PR TITLE
WIP Update Dependencies and Requirements to use `pyproject.toml` 

### PR DESCRIPTION
**Pull request by mfixstsci**
Currently JWQL is using three different methods to track dependencies and build our python environment. It would be nice to shrink this down to one method and keep track of them in the `pyproject.toml`

### CODE REVIEW

**Comment by zacharyburnett** [comment]
Looks good so far! You can likely remove `ruff` and `pytest`, and limit these dependencies only to those imported in the code. I'll draft up a release workflow too

**Comment by mfixstsci** [comment]
@zacharyburnett Thanks for the PR, I went ahead and added the upper and lower bound pins. I also added the minimum python version. I guess the next steps would be to update the CI pipeline to get the tests passing. I am sure there is a nifty and much easier way to get the package installed with the pyproject.toml instead of executing a bunch of bash commands. Ideally the minimum version we would like to support at this point is 3.11 so the testing matrix will need to be updated to include python 3.11 and 3.12.

**Comment by mfixstsci** [comment]
@zacharyburnett so close, looks like something with the mast portal is causing the test failure. I will investigate further tomorrow.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,54 +1,46 @@
...
+name: test
 
-on: [push, pull_request]
+on: 
+  push:
+  pull_request:
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
```
Comment (zacharyburnett) [inline .github/workflows/ci.yml:10]:
```suggestion

defaults:
  run:
    shell: micromamba-shell {0}

```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+name: build
+
+on:
+  release:
+    types: [ released ]
+  pull_request:
+  workflow_dispatch:
+
```
Comment (zacharyburnett) [inline .github/workflows/build.yml:8]:
```suggestion

defaults:
  run:
    shell: micromamba-shell {0}

```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,54 +1,46 @@
...
+        python-version: ["3.11"]
+      fail-fast: false
+    env:
+      PYTHONUNBUFFERED: 1
+    runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v4
-
-      - uses: mamba-org/provision-with-micromamba@v15
+      - uses: actions/setup-python@v5
```
Comment (zacharyburnett) [inline .github/workflows/ci.yml:33]:
```suggestion
      - uses: mamba-org/setup-micromamba@v1
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,54 +1,46 @@
...
-          cache-downloads: true
-
-      - run: pip install -e .[test]
-
-      - run: conda env export
-
-      - uses: supercharge/redis-github-action@1.4.0
+          python-version: ${{ matrix.python-version }}
+          cache: pip
+          cache-dependency-path: pyproject.toml
```
Comment (zacharyburnett) [inline .github/workflows/ci.yml:37]:
```suggestion
          environment-name: jwql-${{ runner.os }}-py${{ matrix.python-version }}
          environment-file: environment.yaml
          create-args: >-
            python=${{ matrix.python-version }}
          init-shell: none
          generate-run-shell: true
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      matrix:
+        python-version:
+          - "3.11"
+      fail-fast: false
+    name: freeze versions (Python ${{ matrix.python-version }})
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
```
Comment (zacharyburnett) [inline .github/workflows/build.yml:26]:
```suggestion
      - uses: mamba-org/setup-micromamba@v1
         with:
          environment-name: jwql-${{ runner.os }}-py${{ matrix.python-version }}
          environment-file: environment.yaml
          create-args: >-
            python=${{ matrix.python-version }}
          init-shell: none
          generate-run-shell: true
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -1,54 +1,46 @@
...
       - run: bandit ./jwql/ -c .bandit
-
   test:
     name: test (Python ${{ matrix.python-version }}, ${{ matrix.os }})
-    runs-on: ${{ matrix.os }}
     strategy:
-      max-parallel: 5
-      fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest]
```
Comment (zacharyburnett) [inline .github/workflows/ci.yml:25]:
```suggestion
        os: [ "ubuntu-latest", "macos-latest" ]
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -1,54 +1,46 @@
...
-        python-version: [3.9, "3.10"]
-
+        python-version: ["3.11"]
```
Comment (zacharyburnett) [inline .github/workflows/ci.yml:26]:
```suggestion
        python-version: [ "3.11" ]
```

**Review by zacharyburnett — CHANGES_REQUESTED** [review]
@mfixstsci use these changes with an `environment.yml` that includes `firefox`:
```yaml
# environment.yml

dependencies:
  - firefox
  - python
```

**Comment by zacharyburnett** [comment]
superseded by #1568

**Pull request closed without merging**